sysboundary/pml.cpp

```diff
@@ -0,0 +1,275 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+/*!\file mpl.cpp
+ * \builds the arrays needed for the PML.
+ */
+
+#include "pml.h"
+#include <cstdlib>
+#include <iostream>
+#include <iomanip> 
+#include <cmath>
+#include <sstream>
+#include <ctime>
+#include <omp.h>
+#include "fsgrid.hpp"
+#include "../definitions.h"
+#include "../common.h"
+#include "../parameters.h"
+#include "../readparameters.h"
+
+
+
+
+bool buildPMLGrids(
+    FsGrid<std::array<Real, fsgrids::pml::N_PML>, 2> &pmlGrid)
+{
+
+   typedef Parameters P;
+
+
+   /*----Get PML Array and Domain Size-----*/
+   std::array<Real, fsgrids::pml::N_PML> *pmlValue;
+   int start = 2;   /*don't judge will change it */
+   std::array<int32_t, 3> pos;
+   const int *pmlDims = &pmlGrid.getLocalSize()[0];
+   const int *globalDims = &pmlGrid.getGlobalSize()[0];
+
+
+
+   /*Read PLM faces to build*/
+   std::vector<std::string> thisSpeciesFaceList;
+   Readparameters::get("PML.face", thisSpeciesFaceList) ;
+   bool Xp, Xm, Yp, Ym, Zp, Zm = false;
+   for (auto &face : thisSpeciesFaceList){
+   if(face == "x+") {Xp = true;}      //set face to true  
+   if(face == "x-") {Xm = true;}      //set face to true
+   if(face == "y+") {Yp = true;}      //set face to true
+   if(face == "y-") {Ym = true;}      //set face to true
+   if(face == "z+") {Zp = true;}      //set face to true
+   if(face == "z-") {Zm = true;}      //set face to true  
+   }   
+
+
+    /*-----Initially set all arrays to one-----*/
+    /*Iterate over domain and set the PML arrays to 1.0 thus not affecting the fieldsolver*/ 
+    #pragma omp parallel for collapse(3)
+    for (int kk = 0; kk < pmlDims[2]; kk++){
+        for (int jj = 0; jj < pmlDims[1]; jj++){
+            for (int ii = 0; ii < pmlDims[0]; ii++){
+
+                pmlValue = pmlGrid.get(ii, jj, kk);
+                pmlValue->at(fsgrids::pml::PGI2) = 1.0;
+                pmlValue->at(fsgrids::pml::PGI3) = 1.0;
+                pmlValue->at(fsgrids::pml::PFI1) = 1.0;
+                pmlValue->at(fsgrids::pml::PFI2) = 1.0;
+                pmlValue->at(fsgrids::pml::PFI3) = 1.0;
+                pmlValue->at(fsgrids::pml::PGJ2) = 1.0;
+                pmlValue->at(fsgrids::pml::PGJ3) = 1.0;
+                pmlValue->at(fsgrids::pml::PFJ1) = 1.0;
+                pmlValue->at(fsgrids::pml::PFJ2) = 1.0;
+                pmlValue->at(fsgrids::pml::PFJ3) = 1.0;
+                pmlValue->at(fsgrids::pml::PGK2) = 1.0;
+                pmlValue->at(fsgrids::pml::PGK3) = 1.0;
+                pmlValue->at(fsgrids::pml::PFK1) = 1.0;
+                pmlValue->at(fsgrids::pml::PFK2) = 1.0;
+                pmlValue->at(fsgrids::pml::PFK3) = 1.0;
+            }
+        }
+    }
+
+
+   Real xnum, xd;
```

Comment:
This is now not thread-safe as these variables will be accessed by all threads. If all the loops are otherwise thread-safe it would anyway be more efficient to start the threaded region `#pragma omp parallel` before the first threaded loop, then have `#pragma omp for collapse(3)` for the loops. This means that the threads are being forked only once and not every loop. Also, then each thread would have its own private copy of these four variables, as you don't want other threads messing with them in between.